qualtran/bloqs/generalized_qsp_test.py

```diff
@@ -251,3 +296,192 @@ def test_generalized_real_qsp_with_symbolic_signal_matrix(degree: int):
     for _ in range(10):
         P = random_qsp_polynomial(degree, random_state=random_state)
         SymbolicGQSP(P).verify()
+
+
+@pytest.mark.parametrize("precision", [1e-5, 1e-7, 1e-10])
+def test_cos_approximation(precision: float):
+    random_state = np.random.RandomState(42)
+
+    for t in [2, 3, 5, 10]:
+        for alpha in [0.5, 1, 2]:
+            bloq = HamiltonianSimulationByGQSP(None, t=t, alpha=alpha, precision=precision)
+
+            P = Polynomial(bloq.approx_cos)
+            theta = 2 * np.pi * random_state.random(1000)
+            e_itheta = np.exp(1j * theta)
+            np.testing.assert_allclose(
+                P(e_itheta) * e_itheta ** (-bloq.degree),
+                np.exp(1j * t * alpha * np.cos(theta)),
+                rtol=precision * 2,
+            )
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("bitsize", [1, 2])
+@pytest.mark.parametrize("t", [2, 3, 5])
+@pytest.mark.parametrize("alpha", [1, 2, 3])
+@pytest.mark.parametrize("precision", [1e-7, 1e-10])
+def test_generalized_qsp_with_exp_cos_approx_on_random_unitaries(
+    bitsize: int, t: float, alpha: float, precision: float
+):
+    random_state = np.random.RandomState(42)
+
+    for _ in range(5):
+        U = RandomGate.create(bitsize, random_state=random_state)
+        gqsp = HamiltonianSimulationByGQSP(None, t=t, alpha=alpha, precision=precision).gqsp
+        P, Q = gqsp.P, gqsp.Q
+
+        check_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=2 * precision
+        )
+        verify_generalized_qsp(U, P, Q, negative_power=len(P) // 2)
+
+
+@frozen
+class RandomPrepareOracle(PrepareOracle):
```

Comment:
Move these newly defined oracles to `bloqs/for_testing/` so they can be reused from elsewhere, if needed.

Reply:
https://github.com/quantumlib/Qualtran/pull/851